eloquentjs/4_2_reverse_array.js

```diff
@@ -0,0 +1,58 @@
+/**
+ * Performs non-mutating array reversal operation
+ * @param {*} input 
+ * @returns 
+ */
+function reverseArray(input) {
+    if (input.constructor.toString().indexOf("Array") === -1) {
+        throw "input must be array";
+    }
+    if (input.length <= 0) {
+        return [];
+    }
+
+    output = [];
+    for (let i = input.length - 1; i >= 0; i--) {
+        output.push(input[i]);
+    }
+    return output;
+}
+
+/**
+ * Performs mutating array reversal operation
+ * @param {*} input 
+ * @returns 
+ */
+function reverseArrayInPlace(input) {
+    if (input.constructor.toString().indexOf("Array") === -1) {
+        throw "input must be array";
+    }
+    if (input.length <= 0) {
+        return [];
+    }
+
+    output = [];
+    while (input.length > 0) {
```

Comment:
This mutates the array, but isn't actually in place. In place means you're not creating another parallel data structure. You can reverse an array in place using only a single variable that stores an element of an array. Try to figure out how to do it that way.